### PR TITLE
Log the average gradient norm

### PR DESCRIPTION
**Pull request by kevinbache**
This PR logs the average L2 norm of the minibatch gradient at each display step.  The average is taken across all the minibatches since the last time the gradient norm was logged. This is useful for helping to track the convergence of a solver; gradient norms should go to zero as we approach a local optima on the loss surface induced by the training dataset.

Tracking the gradient norm incurs a small additional cost per iteration (O(d) where d is the number of parameters in the gradient), but this will be dominated by the gradient computation itself.

### CODE REVIEW

**Comment by kevinbache** [comment]
Also, I forgot to mention that this functionality can be disabled by setting the new solver parameter log_avg_gradient_norm to false (it defaults to true).

**Comment by PiranjaF** [comment]
Great work, thanks! Seems very useful for finding a correct gradient clipping threshold (Pascanu et. al., 'On the diffculty of training recurrent neural networks').

**Comment by shelhamer** [comment]
Closing as handled by `debug_info` https://github.com/BVLC/caffe/blob/master/src/caffe/proto/caffe.proto#L88 in `Net` https://github.com/BVLC/caffe/blob/master/src/caffe/net.cpp#L585-L636

Thanks for considering the monitoring of training.
